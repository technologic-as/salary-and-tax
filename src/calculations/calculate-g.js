import round from './round';

export const calculateG = g => round(99858 * g);

export default calculateG;
