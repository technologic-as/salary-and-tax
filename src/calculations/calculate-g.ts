import round from './round';
import { taxConstants } from '../constants';

export const calculateG = (g: number) => round(taxConstants.oneG * g);

export default calculateG;
