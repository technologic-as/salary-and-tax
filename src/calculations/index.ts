import { defaultSalaryParameters, getSalaryCalculations, SalaryCalculationsState } from './salary';
import { DividendsTaxCalculation, getDividendsTaxCalculations, getTaxCalculations, TaxCalculations } from './tax';
import { ParamProps } from '../constants';

export { defaultSalaryParameters } from './salary';

type Calculations = {
  dividends: DividendsTaxCalculation;
  tax: TaxCalculations;
  afterTotal: number;
  salary: SalaryCalculationsState;
};

export const calculate = (parameters: ParamProps = defaultSalaryParameters): Calculations => {
  const salary = getSalaryCalculations(parameters);
  const tax = getTaxCalculations(salary.income);
  const dividends = getDividendsTaxCalculations(salary.surplus);
  const afterTotal = tax.afterTax + dividends.afterDividendsTax;

  return { salary, tax, dividends, afterTotal };
};

export default calculate;
