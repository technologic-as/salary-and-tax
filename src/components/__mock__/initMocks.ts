const mockFormatMessage = ({ defaultMessage }: { defaultMessage: string }) => defaultMessage;
const mockFormatNumber = (amount: number) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount);

const mockIntl = () => ({
  formatMessage: mockFormatMessage,
  formatDate: jest.fn,
  formatTime: jest.fn,
  formatRelative: jest.fn,
  formatNumber: mockFormatNumber,
  formatPlural: jest.fn,
  formatHTMLMessage: jest.fn,
  now: jest.fn,
});

jest.mock('react-intl', () => ({
  useIntl: mockIntl,
  defineMessages: (msg: unknown) => msg,
}));

jest.mock('react-intl-redux', () => ({
  updateIntl: (a: unknown) => a,
}));

jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/FormGroup', () => 'FormGroup');
jest.mock('@material-ui/core/FormControlLabel', () => 'FormControlLabel');
jest.mock('@material-ui/core/Radio', () => 'Radio');

export {};
