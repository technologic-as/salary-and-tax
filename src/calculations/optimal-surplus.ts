import { calculateG } from './calculate-g';
import { calculate } from './index';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type OptimalSurplus = {
  total: number;
  surplus: any;
  salaryBasis: number;
  dividends: number;
  netIncome: number;
  grossIncome: number;
  salaryAboveSevenG: boolean;
};

export const optimalSurplus = (parameters: ParamProps = defaultSalaryParameters): OptimalSurplus[] => {
  const initialCalculation = calculate(parameters);
  const companyIncome = initialCalculation.salary.companyIncome;
  const increments = parameters.graph.increments;
  const steps = Math.round(companyIncome / increments);

  const sevenG = calculateG(7.1);

  return Array.from(Array(steps).keys())
    .map((i: number) => i * increments)
    .map((amount: number) =>
      calculate({
        ...parameters,
        surplus: {
          amount,
          include: true,
        },
      })
    )
    .map(
      ({
        afterTotal,
        dividends: { afterDividendsTax, surplus },
        salary: { withoutPension, withoutSurplus },
        tax: { afterTax },
      }) => ({
        total: afterTotal,
        surplus,
        dividends: afterDividendsTax,
        grossIncome: withoutPension,
        netIncome: afterTax,
        salaryBasis: withoutSurplus,
        salaryAboveSevenG: withoutPension > sevenG,
      })
    );
};

export default optimalSurplus;
