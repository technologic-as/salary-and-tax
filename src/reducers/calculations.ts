import { FORM_CHANGE, FormChange } from '../actions';
import { calculate, defaultSalaryParameters } from '../calculations';
import { DividendsTaxCalculation, TaxCalculations } from '../calculations/tax';
import { SalaryCalculationsState } from '../calculations/salary';
import { ParamProps } from '../constants';

const initialState: CalculationsState = {
  ...calculate(defaultSalaryParameters),
};

export type CalculationsState = {
  dividends: DividendsTaxCalculation;
  tax: TaxCalculations;
  afterTotal: number;
  salary: SalaryCalculationsState;
};

export const calculationsReducer = (
  state: CalculationsState = initialState,
  action: FormChange = {},
  { parameters }: { parameters?: ParamProps } = {}
): CalculationsState => {
  switch (action.type) {
    case FORM_CHANGE:
      return Object.assign({}, state, {
        ...calculate(parameters),
      });
    default:
      return state;
  }
};

export default calculationsReducer;
