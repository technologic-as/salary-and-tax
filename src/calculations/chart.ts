import { OptimalSurplus, optimalSurplus } from './optimal-surplus';
import { defaultSalaryParameters } from './salary';
import { ParamProps } from '../constants';

export type ChartState = {
  total: (any | number)[][];
  dividends: (any | number)[][];
  netIncome: (any | number)[][];
  grossIncome: (any | number)[][];

  maxTotal: any[][];
  surplus: any[][];
  sevenG: number[][];
};

export const getChart = (parameters: ParamProps = defaultSalaryParameters): ChartState => {
  const data = optimalSurplus(parameters);
  const dividends = data.map((i) => [i.surplus, i.dividends]);
  const surplus = data.map((i) => [i.surplus, i.surplus]);
  const grossIncome = data.map((i) => [i.surplus, i.grossIncome]);
  const netIncome = data.map((i) => [i.surplus, i.netIncome]);
  const total = data.map((i) => [i.surplus, i.total]);
  const closeToSeven = data
    .filter((i) => i.salaryAboveSevenG)
    .reduce((a: OptimalSurplus, b) => (a.salaryBasis < b.salaryBasis ? a : b));
  const sevenG = [[closeToSeven.surplus, closeToSeven.grossIncome]];
  const max = data.reduce((a: OptimalSurplus, b) => (a.total > b.total ? a : b));
  const maxTotal = [[max.surplus, max.total]];

  return {
    sevenG,
    maxTotal,
    grossIncome,
    netIncome,
    surplus,
    dividends,
    total,
  };
};

export default getChart;
