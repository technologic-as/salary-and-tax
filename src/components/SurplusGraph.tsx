import Highcharts from 'highcharts';
import React from 'react';
import { defineMessages, useIntl } from 'react-intl';
import {
  Chart,
  Credits,
  HighchartsChart,
  HighchartsProvider,
  Legend,
  LineSeries,
  ScatterSeries,
  Title,
  Tooltip,
  XAxis,
  YAxis,
} from 'react-jsx-highcharts';
import { connect } from 'react-redux';
import { dividendsClick } from '../actions';
import { Section } from './Ui';
import { State } from '../reducers';
import { ChartState } from '../calculations/chart';
import { Dispatch } from 'redux';

const messages = defineMessages({
  header: {
    id: 'surplus.graph.header',
    defaultMessage: 'Income vs. surplus',
  },
  chartHeader: {
    id: 'surplus.graph.chart.header',
    defaultMessage: 'Income vs. surplus',
  },
  xAxisTitle: {
    id: 'surplus.graph.x.axis.title',
    defaultMessage: 'Surplus',
  },
  yAxisTitle: {
    id: 'surplus.graph.y.axis.title',
    defaultMessage: 'NOK',
  },
  seriesIncomeGross: {
    id: 'surplus.series.income.gross',
    defaultMessage: 'Gross income',
  },
  seriesIncomeNet: {
    id: 'surplus.series.income.net',
    defaultMessage: 'Net income',
  },
  seriesSurplus: {
    id: 'surplus.series.surplus',
    defaultMessage: 'Surplus',
  },
  seriesGrossSevenG: {
    id: 'surplus.series.gross.seven.g',
    defaultMessage: '7.1 Gross income',
  },
  seriesMaxTotal: {
    id: 'surplus.series.max.total',
    defaultMessage: 'Max total',
  },
  seriesDividends: {
    id: 'surplus.series.dividends',
    defaultMessage: 'Dividends',
  },
  seriesTotal: {
    id: 'surplus.series.total',
    defaultMessage: 'Total',
  },
  suffix: {
    id: 'surplus.suffix',
    defaultMessage: 'NOK',
  },
});

type ActionProps = {
  clickDividends: ({ point }: { point: { x: number } }) => void;
};

export const SurplusGraphComponent = ({
  sevenG,
  maxTotal,
  grossIncome,
  netIncome,
  surplus,
  dividends,
  total,
  clickDividends,
}: ChartState & ActionProps) => {
  const { formatMessage } = useIntl();
  return (
    <Section header={formatMessage(messages.header)}>
      <HighchartsProvider Highcharts={Highcharts}>
        <HighchartsChart>
          <Chart />
          <Title>{formatMessage(messages.chartHeader)}</Title>
          <Legend />
          <XAxis>
            <XAxis.Title>{formatMessage(messages.xAxisTitle)}</XAxis.Title>
          </XAxis>
          <YAxis>
            <ScatterSeries name={formatMessage(messages.seriesGrossSevenG)} data={sevenG} />
            <ScatterSeries name={formatMessage(messages.seriesMaxTotal)} data={maxTotal} />
            <YAxis.Title>{formatMessage(messages.yAxisTitle)}</YAxis.Title>
            <LineSeries name={formatMessage(messages.seriesIncomeGross)} data={grossIncome} />
            <LineSeries name={formatMessage(messages.seriesIncomeNet)} data={netIncome} visible={false} />
            <LineSeries name={formatMessage(messages.seriesSurplus)} data={surplus} visible={false} />
            <LineSeries
              name={formatMessage(messages.seriesDividends)}
              data={dividends}
              onClick={(e) => clickDividends(e)}
            />
            <LineSeries name={formatMessage(messages.seriesTotal)} data={total} />
          </YAxis>
          <Tooltip shared valueSuffix={formatMessage(messages.suffix)} />
          <Credits enabled={false} />
        </HighchartsChart>
      </HighchartsProvider>
    </Section>
  );
};

const mapStateToProps = ({
  graph: { sevenG, maxTotal, grossIncome, netIncome, surplus, dividends, total },
}: State) => ({
  grossIncome,
  netIncome,
  surplus,
  dividends,
  total,
  sevenG,
  maxTotal,
});

const mapDispatchToProps = (dispatch: Dispatch): ActionProps => ({
  clickDividends: ({ point }) => dividendsClick(point)(dispatch),
});

export const SurplusGraph = connect(mapStateToProps, mapDispatchToProps)(SurplusGraphComponent);

export default SurplusGraph;
