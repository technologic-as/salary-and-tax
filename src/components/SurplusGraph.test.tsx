import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import mockComponent from 'identity-obj-proxy';
import React from 'react';
import './__mock__/initMocks';
import { defaultSalaryParameters } from '../calculations';
import { getChart } from '../calculations/chart';
import renderer from 'react-test-renderer';
import { SurplusGraphComponent } from './SurplusGraph';

jest.mock('./Ui', () => mockComponent);
jest.mock('highcharts', () => ({}));

jest.mock('react-jsx-highcharts', () => {
  const axisMock = ({ children }: { children: unknown }) => <>{children}</>;
  axisMock.Title = 'Title';
  return {
    XAxis: axisMock,
    YAxis: axisMock,
    HighchartsProvider: 'HighchartsProvider',
    HighchartsChart: 'HighchartsChart',
    Chart: 'Chart',
    Credits: 'Credits',
    Title: 'Title',
    Legend: 'Legend',
    ScatterSeries: 'ScatterSeries',
    LineSeries: 'LineSeries',
    Tooltip: 'Tooltip',
  };
});

beforeAll(() => {
  Enzyme.configure({ adapter: new Adapter() });
});

describe('SurplusGraph', () => {
  it('should render', () => {
    const { sevenG, maxTotal, grossIncome, netIncome, surplus, dividends, total } = getChart({
      ...defaultSalaryParameters,
      graph: { increments: 100000 },
    });
    const tree = renderer
      .create(
        <SurplusGraphComponent
          sevenG={sevenG}
          maxTotal={maxTotal}
          grossIncome={grossIncome}
          netIncome={netIncome}
          surplus={surplus}
          dividends={dividends}
          total={total}
          clickDividends={jest.fn()}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
