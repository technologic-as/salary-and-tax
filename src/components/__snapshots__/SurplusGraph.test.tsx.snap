// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurplusGraph should render 1`] = `
<Section
  header="Income vs. surplus"
>
  <HighchartsProvider
    Highcharts={Object {}}
  >
    <HighchartsChart>
      <Chart />
      <Title>
        Income vs. surplus
      </Title>
      <Legend />
      <Title>
        Surplus
      </Title>
      <ScatterSeries
        data={
          Array [
            Array [
              600000,
              855529,
            ],
          ]
        }
        name="7.1 Gross income"
      />
      <ScatterSeries
        data={
          Array [
            Array [
              900000,
              915516,
            ],
          ]
        }
        name="Max total"
      />
      <Title>
        NOK
      </Title>
      <LineSeries
        data={
          Array [
            Array [
              0,
              1315701,
            ],
            Array [
              100000,
              1238965,
            ],
            Array [
              200000,
              1162277,
            ],
            Array [
              300000,
              1085590,
            ],
            Array [
              400000,
              1008903,
            ],
            Array [
              500000,
              932216,
            ],
            Array [
              600000,
              855529,
            ],
            Array [
              700000,
              778841,
            ],
            Array [
              800000,
              702155,
            ],
            Array [
              900000,
              625467,
            ],
            Array [
              1000000,
              548780,
            ],
            Array [
              1100000,
              472093,
            ],
            Array [
              1200000,
              395406,
            ],
            Array [
              1300000,
              318719,
            ],
            Array [
              1400000,
              242031,
            ],
            Array [
              1500000,
              165344,
            ],
            Array [
              1600000,
              88192,
            ],
          ]
        }
        name="Gross income"
      />
      <LineSeries
        data={
          Array [
            Array [
              0,
              842620,
            ],
            Array [
              100000,
              801490,
            ],
            Array [
              200000,
              760386,
            ],
            Array [
              300000,
              719282,
            ],
            Array [
              400000,
              678177,
            ],
            Array [
              500000,
              635963,
            ],
            Array [
              600000,
              592558,
            ],
            Array [
              700000,
              549152,
            ],
            Array [
              800000,
              505748,
            ],
            Array [
              900000,
              460620,
            ],
            Array [
              1000000,
              410006,
            ],
            Array [
              1100000,
              359393,
            ],
            Array [
              1200000,
              308779,
            ],
            Array [
              1300000,
              258166,
            ],
            Array [
              1400000,
              206957,
            ],
            Array [
              1500000,
              146698,
            ],
            Array [
              1600000,
              84883,
            ],
          ]
        }
        name="Net income"
        visible={false}
      />
      <LineSeries
        data={
          Array [
            Array [
              0,
              0,
            ],
            Array [
              100000,
              100000,
            ],
            Array [
              200000,
              200000,
            ],
            Array [
              300000,
              300000,
            ],
            Array [
              400000,
              400000,
            ],
            Array [
              500000,
              500000,
            ],
            Array [
              600000,
              600000,
            ],
            Array [
              700000,
              700000,
            ],
            Array [
              800000,
              800000,
            ],
            Array [
              900000,
              900000,
            ],
            Array [
              1000000,
              1000000,
            ],
            Array [
              1100000,
              1100000,
            ],
            Array [
              1200000,
              1200000,
            ],
            Array [
              1300000,
              1300000,
            ],
            Array [
              1400000,
              1400000,
            ],
            Array [
              1500000,
              1500000,
            ],
            Array [
              1600000,
              1600000,
            ],
          ]
        }
        name="Surplus"
        visible={false}
      />
      <LineSeries
        data={
          Array [
            Array [
              0,
              0,
            ],
            Array [
              100000,
              50544,
            ],
            Array [
              200000,
              101088,
            ],
            Array [
              300000,
              151632,
            ],
            Array [
              400000,
              202176,
            ],
            Array [
              500000,
              252720,
            ],
            Array [
              600000,
              303264,
            ],
            Array [
              700000,
              353808,
            ],
            Array [
              800000,
              404352,
            ],
            Array [
              900000,
              454896,
            ],
            Array [
              1000000,
              505440,
            ],
            Array [
              1100000,
              555984,
            ],
            Array [
              1200000,
              606528,
            ],
            Array [
              1300000,
              657072,
            ],
            Array [
              1400000,
              707616,
            ],
            Array [
              1500000,
              758160,
            ],
            Array [
              1600000,
              808704,
            ],
          ]
        }
        name="Dividends"
        onClick={[Function]}
      />
      <LineSeries
        data={
          Array [
            Array [
              0,
              842620,
            ],
            Array [
              100000,
              852034,
            ],
            Array [
              200000,
              861474,
            ],
            Array [
              300000,
              870914,
            ],
            Array [
              400000,
              880353,
            ],
            Array [
              500000,
              888683,
            ],
            Array [
              600000,
              895822,
            ],
            Array [
              700000,
              902960,
            ],
            Array [
              800000,
              910100,
            ],
            Array [
              900000,
              915516,
            ],
            Array [
              1000000,
              915446,
            ],
            Array [
              1100000,
              915377,
            ],
            Array [
              1200000,
              915307,
            ],
            Array [
              1300000,
              915238,
            ],
            Array [
              1400000,
              914573,
            ],
            Array [
              1500000,
              904858,
            ],
            Array [
              1600000,
              893587,
            ],
          ]
        }
        name="Total"
      />
      <Tooltip
        shared={true}
        valueSuffix="NOK"
      />
      <Credits
        enabled={false}
      />
    </HighchartsChart>
  </HighchartsProvider>
</Section>
`;
